app/controllers/application_controller.rb

```diff
@@ -2,9 +2,156 @@
 
 class ApplicationController < Sinatra::Base
 
+  helpers SessionHelpers
+  helpers UserHelpers
+  helpers TweetHelpers
+
+  enable :sessions
+  set :session_secret, "secret"
+  use Rack::Flash
+
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
   end
 
-end
\ No newline at end of file
+  get '/' do
+    erb :index
+  end
+
+  get '/signup' do
+    redirect to "/tweets" if is_logged_in?
+
+    erb :'/users/create_user'
+  end
+
+  get '/login' do
+    redirect to "/tweets" if is_logged_in?
+
+    erb :'users/login'
+  end
+
+  get '/logout' do
+    session.clear if is_logged_in?
+    redirect to "/login"
+  end
+
+  get '/tweets/new' do
+    if is_logged_in?
+      erb :"/tweets/create_tweet"
+    else
+      flash[:notice] = "You must new logged in to tweet!"
+      redirect to "/login"
+    end
+  end
+
+  post '/tweets' do
+    if is_logged_in?
+      if validate_new_tweet_params(params)
+        tweet = Tweet.create(content: params[:content],
+                             user_id: current_user.id)
+
+        redirect to "/users/#{current_user.username}"
+      else
+        flash[:notice] = "Tweets may not be empty."
+        redirect to "/tweets/new"
+      end
+    else
+      flash[:notice] = "You must be logged in to tweet"
+      redirect to "/login"
+    end
+  end
+
+  get '/tweets/:id/edit' do
+    @tweet = Tweet.find_by(id: params[:id])
+
+    if @tweet && is_logged_in?
+      if current_user.id == @tweet.user_id
+        erb :'/tweets/edit'
+      else
+        redirect to "/tweets"
+      end
+    else
+      redirect to "/login"
+    end
+  end
+
+  patch '/tweets/:id' do
+    @tweet = Tweet.find(params[:id])
+
+    if params[:content] == "" && is_logged_in?
+      redirect to "/tweets/#{@tweet.id}/edit"
+    elsif @tweet && current_user.id == @tweet.user_id
+      @tweet.update(content: params[:content])
+      redirect to "/tweets/#{@tweet.id}"
+    else
+      redirect to "/login"
+    end
+  end
+
+  delete '/tweets/:id' do
+    if is_logged_in?
+      tweet = Tweet.find_by(id: params[:id])
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.